engine/unit_test.go

```diff
@@ -19,13 +20,21 @@ func TestReadyDone(t *testing.T) {
 // Test that if a function is run by LaunchPeriodically and
 // takes longer than the interval, the next call will be blocked
 func TestLaunchPeriod(t *testing.T) {
+	lock := sync.Mutex{}
+
 	u := engine.NewUnit()
 	unittest.RequireCloseBefore(t, u.Ready(), time.Second, "ready did not close")
 	logs := make([]string, 0)
 	u.LaunchPeriodically(func() {
+		lock.Lock()
 		logs = append(logs, "running")
+		lock.Unlock()
```

Comment:
I guess @SaveTheRbtz must have found this issue with the race detector, right?

However, when I ran the test with `-race` I ran into the following error. (Test passes when ran it without `-race`).

It's the same for both master and your branch. I wonder if you ran into the same issue?

```
▸ go test -race . -v -run=TestLaunchPeriod  -tags=relic -count=1
fatal error: checkptr: pointer arithmetic result points to invalid allocation

goroutine 1 [running, locked to thread]:
runtime.throw({0x4f5a7b4?, 0x406e879?})
        /usr/local/opt/go/libexec/src/runtime/panic.go:992 +0x71 fp=0xc00029f668 sp=0xc00029f638 pc=0x4037d71
runtime.checkptrArithmetic(0x25?, {0x0, 0x0, 0x57cf620?})
        /usr/local/opt/go/libexec/src/runtime/checkptr.go:69 +0xaa fp=0xc00029f698 sp=0xc00029f668 pc=0x4008e8a
github.com/onflow/flow-go/crypto/hash.(*storageBuf).asBytes(...)
        /Users/leo/gopath/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/hash/xor_unaligned.go:48
github.com/onflow/flow-go/crypto/hash.(*spongeState).appendBuf(...)
        /Users/leo/gopath/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/hash/keccak.go:122
github.com/onflow/flow-go/crypto/hash.(*spongeState).write(0xc00029f740, {0xc00003d2c0, 0x25, 0x25})
        /Users/leo/gopath/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/hash/keccak.go:160 +0x2a5 fp=0xc00029f718 sp=0xc00029f698 pc=0x43248e5
github.com/onflow/flow-go/crypto/hash.ComputeSHA3_256(0xc00029f8f0, {0xc00003d2c0, 0x25, 0x25})
        /Users/leo/gopath/pkg/mod/github.com/onflow/flow-go/crypto@v0.24.3/hash/sha3.go:48 +0xcd fp=0xc00029f8d0 sp=0xc00029f718 pc=0x432646d
github.com/onflow/flow-go/model/flow.MakeID({0x4e93320, 0xc000310f80})
        /Users/leo/flow-go/model/flow/identifier.go:119 +0x66 fp=0xc00029f920 sp=0xc00029f8d0 pc=0x45c3766
github.com/onflow/flow-go/model/cluster.Payload.Hash(...)
        /Users/leo/flow-go/model/cluster/payload.go:53
github.com/onflow/flow-go/state/cluster.init()
        /Users/leo/flow-go/state/cluster/root_block.go:18 +0x22d fp=0xc00029fac0 sp=0xc00029f920 pc=0x4bf6f0d
runtime.doInit(0x57752c0)
        /usr/local/opt/go/libexec/src/runtime/proc.go:6222 +0x126 fp=0xc00029fbf0 sp=0xc00029fac0 pc=0x4047786
runtime.doInit(0x5783740)
        /usr/local/opt/go/libexec/src/runtime/proc.go:6199 +0x71 fp=0xc00029fd20 sp=0xc00029fbf0 pc=0x40476d1
runtime.doInit(0x577ae80)
        /usr/local/opt/go/libexec/src/runtime/proc.go:6199 +0x71 fp=0xc00029fe50 sp=0xc00029fd20 pc=0x40476d1
runtime.doInit(0x5777800)
        /usr/local/opt/go/libexec/src/runtime/proc.go:6199 +0x71 fp=0xc00029ff80 sp=0xc00029fe50 pc=0x40476d1
runtime.main()
        /usr/local/opt/go/libexec/src/runtime/proc.go:233 +0x1d3 fp=0xc00029ffe0 sp=0xc00029ff80 pc=0x403a413
runtime.goexit()
        /usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc00029ffe8 sp=0xc00029ffe0 pc=0x406d141
FAIL    github.com/onflow/flow-go/engine        0.389s
FAIL

▸ go version
go version go1.18 darwin/amd64
```

Reply:
You can try applying this diff https://github.com/onflow/flow-go/pull/2293/commits/1288c1dbcfa19c4249cdc0ae88078e7f4112b4d5 to the hasher inside `$GOPATH/pkg/mod/github.com/onflow/flow-go` (IIUC `crypto` package is installed separately)

Reply:
I'm still running into this error even with [the diff](https://github.com/onflow/flow-go/commit/1288c1dbcfa19c4249cdc0ae88078e7f4112b4d5) on top of latest master or your branch.

Tried on MacOS with golang v1.17 and 1.18.

Reply:
Interesting, do you see it also applied to files in `$GOPATH/pkg/mod/github.com/onflow/flow-go`?